test/system/src/test/java/io/pravega/test/system/AbstractReadWriteTest.java

```diff
@@ -225,6 +225,7 @@ public void cancelAllPendingWork() {
                     String uniqueRoutingKey = transaction.getTxnId().toString();
                     long seqNumber = 0;
                     for (int j = 1; j <= NUM_EVENTS_PER_TRANSACTION; j++) {
+                        Exceptions.handleInterrupted(() -> Thread.sleep(100));
```

Comment:
This is hardcoding how we space out transactions. What about throttling using the number of outstanding transactions? We are keeping the futures for transactions in a transactions status future list, so we can change this loop a bit to check how many transactions are uncommitted before creating a new one. We could also make the threshold on the number of outstanding transactions configurable on a per test case basis.

Reply:
While your comment makes sense, I would suggest to keep this kind of throttling due to the following:
1) The problem we are observing may not about transactions being accumulated in `COMMITTING` state right from the beginning. For instance, we have observed in the first stages of the test that writers were able to commit tens of transactions per second. The problem is that, as more data and metadata is getting accumulated, it makes some HDFS operations to slow down progressively. This has consequences later on, specially impacting some tests that perform a quick succession of failovers and recoveries that are assumed to meet some recovery times. Therefore, the problem can be created when transactions are being committed fast, and when we reach a point in which we could detect outstanding transactions being accumulated it may be too late for throttling to be effective. Solving this calls for a hard throttling rate for the whole lifecycle of transactional writers. 
2) It is consistent with the kind of throttling we are already enforcing for regular writers: https://github.com/pravega/pravega/blob/f7ef2488e7ca0a5fd16502d8d67d4c3821436930/test/system/src/test/java/io/pravega/test/system/AbstractReadWriteTest.java#L182
3) I have also tried to place the sleep "outside" the loop, leading to a single transactional writer to  constrain its write throughput to at most 10 transactions/second. In this case, we have also experienced several tests failures in HDFS, which indicates an upper bound on the load we can handle in system tests with HDFS.